lib/api/src/modules/stories.ts

```diff
@@ -364,7 +364,19 @@ Did you create a path that uses the separator char accidentally, such as 'Vue <d
       const kind = storyId.split('--', 2)[0];
       selectStory(toId(kind, story));
     } else {
-      selectStory(toId(kindOrId, story));
+      const id = toId(kindOrId, story);
+      if (storiesHash[id]) {
+        selectStory(id);
+      } else {
+        // Support legacy API with component permalinks, where kind is `x/y` but permalink is 'z'
+        const k = storiesHash[sanitize(kindOrId)];
+        if (k && k.children) {
+          const foundId = k.children.find(childId => storiesHash[childId].name === story);
```

Comment:
Would it be better to do:

```js
// This kind has a custom id, look up the story using it
if (k && k.id !== sanitize(kindOrId))
  return selectStory(k.id, story)
```

PS prefer `kind` to `k` if the code is more than one line.

Reply:
1) Unfortunately k.id is sanitize(kindOrId)
2) `kind` was already in scope

Reply:
Is there nothing stored on the kind in the store to indicate its custom id? Should we change that?

Reply:
I feel like this line of code is pretty dodgy and we'll regret it in the long run

Reply:
I agree that the entire setup is lacking integrity, but I don't have a good idea for a better setup. Any suggestions?

Reply:
My suggestion would be to store the `componentId` on the data structure for a kind (I guess in the story store?), then you can do a line like I did above:

```js
if (k && k.componentId !== sanitize(kindOrId)) return selectStory(k.id, story)
```

Reply:
Where would this get passed in:

```js
storiesOf({ kind, componentId }, module).add(...);

storiesOf(kind, module).setComponentId(componentId);
```

???

Reply:
Oh, so currently we don't store `componentId` at all, we just change the individual stories `id`s based on it?

Reply:
I would suggest `setComponentId()`

Reply:
@tmeasday there's no intermediate data structure in `story_store` to hold this AFAIK. i feel like this is all out of scope.

Reply:
OK, fair enough.